File: catkin_ws/src/camera_lidar_fusion/scripts/camera_lidar_fusion_node.py
#!/usr/bin/env python3

import rospy
import numpy as np
import cv2
from sensor_msgs.msg import PointCloud2, Image, CameraInfo
from geometry_msgs.msg import TransformStamped
import tf2_ros
import tf2_geometry_msgs
from cv_bridge import CvBridge
import message_filters
from message_filters import ApproximateTimeSynchronizer, Subscriber
import pcl
import pcl_helper

class CameraLidarFusionPython:
    def __init__(self):
        rospy.init_node('camera_lidar_fusion_python', anonymous=True)
        
        # Initialize CV bridge
        self.bridge = CvBridge()
        
        # Initialize TF
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)
        self.tf_broadcaster = tf2_ros.TransformBroadcaster()
        
        # Publishers
        self.fused_cloud_pub = rospy.Publisher('/fused/cloud', PointCloud2, queue_size=1)
        self.colored_cloud_pub = rospy.Publisher('/fused/colored_cloud', PointCloud2, queue_size=1)
        self.debug_image_pub = rospy.Publisher('/fused/debug_image', Image, queue_size=1)
        
        # Subscribers
        cloud_sub = Subscriber('/unilidar/cloud', PointCloud2)
        image_sub = Subscriber('/zed/left/image_rect_color', Image)
        camera_info_sub = Subscriber('/zed/left/camera_info', CameraInfo)
        
        # Synchronizer
        self.ts = ApproximateTimeSynchronizer([cloud_sub, image_sub, camera_info_sub], queue_size=10, slop=0.1)
        self.ts.registerCallback(self.callback)
        
        # Calibration parameters (you'll need to calibrate these)
        self.camera_matrix = np.array([[1000.0, 0, 640.0],
                                      [0, 1000.0, 480.0],
                                      [0, 0, 1]], dtype=np.float64)
        self.dist_coeffs = np.zeros(5)
        
        # LiDAR to camera transform (you'll need to calibrate these)
        self.rotation_matrix = np.eye(3)
        self.translation_vector = np.array([0.1, 0, 0])  # 10cm offset in X
        
        # Processing parameters
        self.voxel_leaf_size = rospy.get_param('~voxel_leaf_size', 0.03)
        self.z_min = rospy.get_param('~z_min', -0.2)
        self.z_max = rospy.get_param('~z_max', 2.5)
        self.sor_mean_k = rospy.get_param('~sor_mean_k', 20)
        self.sor_std_dev = rospy.get_param('~sor_std_dev', 1.0)
        
        # Publish static transform
        self.publish_static_transform()
        
        rospy.loginfo("Camera-LiDAR fusion Python node initialized")
    
    def publish_static_transform(self):
        """Publish static transform from LiDAR to camera"""
        transform = TransformStamped()
        transform.header.stamp = rospy.Time.now()
        transform.header.frame_id = "camera_link"
        transform.child_frame_id = "lidar_link"
        
        transform.transform.translation.x = self.translation_vector[0]
        transform.transform.translation.y = self.translation_vector[1]
        transform.transform.translation.z = self.translation_vector[2]
        
        # Convert rotation matrix to quaternion
        from tf.transformations import quaternion_from_matrix
        rotation_matrix_4x4 = np.eye(4)
        rotation_matrix_4x4[:3, :3] = self.rotation_matrix
        quat = quaternion_from_matrix(rotation_matrix_4x4)
        
        transform.transform.rotation.x = quat[0]
        transform.transform.rotation.y = quat[1]
        transform.transform.rotation.z = quat[2]
        transform.transform.rotation.w = quat[3]
        
        self.tf_broadcaster.sendTransform(transform)
    
    def filter_point_cloud(self, cloud):
        """Apply filters to point cloud"""
        # Convert to PCL format
        pcl_cloud = pcl_helper.ros_to_pcl(cloud)
        
        # Voxel downsampling
        voxel_filter = pcl_cloud.make_voxel_grid_filter()
        voxel_filter.set_leaf_size(self.voxel_leaf_size, self.voxel_leaf_size, self.voxel_leaf_size)
        filtered_cloud = voxel_filter.filter()
        
        # Z-axis filtering
        pass_filter = filtered_cloud.make_passthrough_filter()
        pass_filter.set_filter_field_name('z')
        pass_filter.set_filter_limits(self.z_min, self.z_max)
        filtered_cloud = pass_filter.filter()
        
        # Statistical outlier removal
        sor_filter = filtered_cloud.make_statistical_outlier_filter()
        sor_filter.set_mean_k(self.sor_mean_k)
        sor_filter.set_std_dev_mul_thresh(self.sor_std_dev)
        filtered_cloud = sor_filter.filter()
        
        return filtered_cloud
    
    def color_point_cloud(self, cloud, image_msg, camera_info_msg):
        """Color point cloud with camera data"""
        try:
            # Convert ROS image to OpenCV
            cv_image = self.bridge.imgmsg_to_cv2(image_msg, "bgr8")
        except Exception as e:
            rospy.logerr(f"Error converting image: {e}")
            return
        
        # Convert PCL cloud to numpy array
        cloud_array = pcl_helper.pcl_to_ros(cloud)
        points = np.array(cloud_array.data).reshape(-1, 3)
        
        # Create colored points array
        colored_points = np.zeros((len(points), 6))  # x, y, z, r, g, b
        colored_points[:, :3] = points
        
        debug_image = cv_image.copy()
        
        for i, point in enumerate(points):
            # Transform point from LiDAR to camera coordinates
            lidar_point = np.array([point[0], point[1], point[2]])
            camera_point = self.rotation_matrix @ lidar_point + self.translation_vector
            
            # Project to image plane
            if camera_point[2] > 0:  # Point is in front of camera
                image_point = self.camera_matrix @ camera_point
                u = image_point[0] / image_point[2]
                v = image_point[1] / image_point[2]
                
                # Check if point projects within image bounds
                if 0 <= u < cv_image.shape[1] and 0 <= v < cv_image.shape[0]:
                    color = cv_image[int(v), int(u)]
                    
                    colored_points[i, 3] = color[2]  # R
                    colored_points[i, 4] = color[1]  # G
                    colored_points[i, 5] = color[0]  # B
                    
                    # Draw point on debug image
                    cv2.circle(debug_image, (int(u), int(v)), 2, (0, 255, 0), -1)
                else:
                    # Point outside image - use default color
                    colored_points[i, 3:6] = [128, 128, 128]
            else:
                # Point behind camera - use default color
                colored_points[i, 3:6] = [128, 128, 128]
        
        # Create colored point cloud message
        colored_cloud_msg = pcl_helper.array_to_pointcloud2(colored_points, cloud.header)
        colored_cloud_msg.header = cloud.header
        self.colored_cloud_pub.publish(colored_cloud_msg)
        
        # Publish debug image
        try:
            debug_image_msg = self.bridge.cv2_to_imgmsg(debug_image, "bgr8")
            debug_image_msg.header = image_msg.header
            self.debug_image_pub.publish(debug_image_msg)
        except Exception as e:
            rospy.logerr(f"Error publishing debug image: {e}")
    
    def callback(self, cloud_msg, image_msg, camera_info_msg):
        """Main callback function"""
        try:
            # Filter point cloud
            filtered_cloud = self.filter_point_cloud(cloud_msg)
            
            # Publish filtered cloud
            filtered_cloud_msg = pcl_helper.pcl_to_ros(filtered_cloud)
            filtered_cloud_msg.header = cloud_msg.header
            self.fused_cloud_pub.publish(filtered_cloud_msg)
            
            # Color point cloud
            self.color_point_cloud(filtered_cloud, image_msg, camera_info_msg)
            
            rospy.loginfo(f"Processed cloud with {len(filtered_cloud)} points")
            
        except Exception as e:
            rospy.logerr(f"Error in callback: {e}")

if __name__ == '__main__':
    try:
        fusion_node = CameraLidarFusionPython()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

File: catkin_ws/src/camera_lidar_fusion/src/camera_lidar_fusion_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/PointCloud2.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/CameraInfo.h>
#include <geometry_msgs/TransformStamped.h>
#include <tf2_ros/transform_broadcaster.h>
#include <tf2_ros/static_transform_broadcaster.h>
#include <tf2/LinearMath/Transform.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>
#include <pcl_conversions/pcl_conversions.h>
#include <pcl/filters/voxel_grid.h>
#include <pcl/filters/passthrough.h>
#include <pcl/filters/statistical_outlier_removal.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <opencv2/calib3d.hpp>
#include <message_filters/subscriber.h>
#include <message_filters/synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>

class CameraLidarFusion {
private:
    ros::NodeHandle nh_;
    ros::NodeHandle private_nh_;
    
    // Publishers
    ros::Publisher fused_cloud_pub_;
    ros::Publisher colored_cloud_pub_;
    ros::Publisher debug_image_pub_;
    
    // Subscribers
    message_filters::Subscriber<sensor_msgs::PointCloud2> cloud_sub_;
    message_filters::Subscriber<sensor_msgs::Image> image_sub_;
    message_filters::Subscriber<sensor_msgs::CameraInfo> camera_info_sub_;
    
    // Synchronizer
    typedef message_filters::sync_policies::ApproximateTime<
        sensor_msgs::PointCloud2, sensor_msgs::Image, sensor_msgs::CameraInfo> SyncPolicy;
    typedef message_filters::Synchronizer<SyncPolicy> Sync;
    boost::shared_ptr<Sync> sync_;
    
    // TF
    tf2_ros::TransformBroadcaster tf_broadcaster_;
    tf2_ros::StaticTransformBroadcaster static_tf_broadcaster_;
    
    // Calibration parameters
    cv::Mat camera_matrix_;
    cv::Mat dist_coeffs_;
    cv::Mat rotation_matrix_;
    cv::Mat translation_vector_;
    
    // Processing parameters
    double voxel_leaf_size_;
    double z_min_, z_max_;
    int sor_mean_k_;
    double sor_std_dev_;
    
    // Point cloud processing
    pcl::VoxelGrid<pcl::PointXYZ> voxel_filter_;
    pcl::PassThrough<pcl::PointXYZ> pass_through_filter_;
    pcl::StatisticalOutlierRemoval<pcl::PointXYZ> sor_filter_;

public:
    CameraLidarFusion() : private_nh_("~") {
        // Initialize parameters
        private_nh_.param("voxel_leaf_size", voxel_leaf_size_, 0.03);
        private_nh_.param("z_min", z_min_, -0.2);
        private_nh_.param("z_max", z_max_, 2.5);
        private_nh_.param("sor_mean_k", sor_mean_k_, 20);
        private_nh_.param("sor_std_dev", sor_std_dev_, 1.0);
        
        // Initialize publishers
        fused_cloud_pub_ = nh_.advertise<sensor_msgs::PointCloud2>("/fused/cloud", 1);
        colored_cloud_pub_ = nh_.advertise<sensor_msgs::PointCloud2>("/fused/colored_cloud", 1);
        debug_image_pub_ = nh_.advertise<sensor_msgs::Image>("/fused/debug_image", 1);
        
        // Initialize subscribers
        cloud_sub_.subscribe(nh_, "/unilidar/cloud", 1);
        image_sub_.subscribe(nh_, "/zed/left/image_rect_color", 1);
        camera_info_sub_.subscribe(nh_, "/zed/left/camera_info", 1);
        
        // Initialize synchronizer
        sync_.reset(new Sync(SyncPolicy(10), cloud_sub_, image_sub_, camera_info_sub_));
        sync_->registerCallback(boost::bind(&CameraLidarFusion::callback, this, _1, _2, _3));
        
        // Initialize filters
        voxel_filter_.setLeafSize(voxel_leaf_size_, voxel_leaf_size_, voxel_leaf_size_);
        pass_through_filter_.setFilterFieldName("z");
        pass_through_filter_.setFilterLimits(z_min_, z_max_);
        sor_filter_.setMeanK(sor_mean_k_);
        sor_filter_.setStddevMulThresh(sor_std_dev_);
        
        // Load calibration (you'll need to provide these values)
        loadCalibration();
        
        // Publish static transform from LiDAR to camera
        publishStaticTransform();
        
        ROS_INFO("Camera-LiDAR fusion node initialized");
    }
    
    void loadCalibration() {
        // These values should be obtained from calibration
        // For now, using default values - you'll need to calibrate your setup
        camera_matrix_ = cv::Mat::eye(3, 3, CV_64F);
        camera_matrix_.at<double>(0, 0) = 1000.0; // fx
        camera_matrix_.at<double>(1, 1) = 1000.0; // fy
        camera_matrix_.at<double>(0, 2) = 640.0;  // cx
        camera_matrix_.at<double>(1, 2) = 480.0;  // cy
        
        dist_coeffs_ = cv::Mat::zeros(5, 1, CV_64F);
        
        // Rotation and translation from LiDAR to camera
        rotation_matrix_ = cv::Mat::eye(3, 3, CV_64F);
        translation_vector_ = cv::Mat::zeros(3, 1, CV_64F);
        translation_vector_.at<double>(0, 0) = 0.1; // 10cm offset in X
    }
    
    void publishStaticTransform() {
        geometry_msgs::TransformStamped transform;
        transform.header.stamp = ros::Time::now();
        transform.header.frame_id = "camera_link";
        transform.child_frame_id = "lidar_link";
        
        // Set transform based on calibration
        transform.transform.translation.x = translation_vector_.at<double>(0, 0);
        transform.transform.translation.y = translation_vector_.at<double>(1, 0);
        transform.transform.translation.z = translation_vector_.at<double>(2, 0);
        
        // Convert rotation matrix to quaternion
        cv::Mat R = rotation_matrix_;
        tf2::Matrix3x3 tf_rotation(
            R.at<double>(0, 0), R.at<double>(0, 1), R.at<double>(0, 2),
            R.at<double>(1, 0), R.at<double>(1, 1), R.at<double>(1, 2),
            R.at<double>(2, 0), R.at<double>(2, 1), R.at<double>(2, 2)
        );
        tf2::Quaternion q;
        tf_rotation.getRotation(q);
        
        transform.transform.rotation.x = q.x();
        transform.transform.rotation.y = q.y();
        transform.transform.rotation.z = q.z();
        transform.transform.rotation.w = q.w();
        
        static_tf_broadcaster_.sendTransform(transform);
    }
    
    void callback(const sensor_msgs::PointCloud2::ConstPtr& cloud_msg,
                  const sensor_msgs::Image::ConstPtr& image_msg,
                  const sensor_msgs::CameraInfo::ConstPtr& camera_info_msg) {
        
        // Process point cloud
        pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
        pcl::fromROSMsg(*cloud_msg, *cloud);
        
        // Apply filters
        pcl::PointCloud<pcl::PointXYZ>::Ptr filtered_cloud(new pcl::PointCloud<pcl::PointXYZ>);
        
        // Voxel downsampling
        voxel_filter_.setInputCloud(cloud);
        voxel_filter_.filter(*filtered_cloud);
        
        // Z-axis filtering
        pass_through_filter_.setInputCloud(filtered_cloud);
        pass_through_filter_.filter(*filtered_cloud);
        
        // Statistical outlier removal
        sor_filter_.setInputCloud(filtered_cloud);
        sor_filter_.filter(*filtered_cloud);
        
        // Publish filtered cloud
        sensor_msgs::PointCloud2 filtered_msg;
        pcl::toROSMsg(*filtered_cloud, filtered_msg);
        filtered_msg.header = cloud_msg->header;
        fused_cloud_pub_.publish(filtered_msg);
        
        // Color the point cloud with camera data
        colorPointCloud(filtered_cloud, image_msg, camera_info_msg);
        
        ROS_INFO("Processed cloud with %lu points", filtered_cloud->size());
    }
    
    void colorPointCloud(const pcl::PointCloud<pcl::PointXYZ>::Ptr& cloud,
                        const sensor_msgs::Image::ConstPtr& image_msg,
                        const sensor_msgs::CameraInfo::ConstPtr& camera_info_msg) {
        
        // Convert ROS image to OpenCV
        cv_bridge::CvImagePtr cv_ptr;
        try {
            cv_ptr = cv_bridge::toCvCopy(image_msg, sensor_msgs::image_encodings::BGR8);
        } catch (cv_bridge::Exception& e) {
            ROS_ERROR("cv_bridge exception: %s", e.what());
            return;
        }
        
        // Create colored point cloud
        pcl::PointCloud<pcl::PointXYZRGB>::Ptr colored_cloud(new pcl::PointCloud<pcl::PointXYZRGB>);
        colored_cloud->points.resize(cloud->size());
        
        cv::Mat debug_image = cv_ptr->image.clone();
        
        for (size_t i = 0; i < cloud->size(); ++i) {
            const pcl::PointXYZ& pt = cloud->points[i];
            
            // Transform point from LiDAR to camera coordinates
            cv::Mat lidar_point(3, 1, CV_64F);
            lidar_point.at<double>(0, 0) = pt.x;
            lidar_point.at<double>(1, 0) = pt.y;
            lidar_point.at<double>(2, 0) = pt.z;
            
            cv::Mat camera_point = rotation_matrix_ * lidar_point + translation_vector_;
            
            // Project to image plane
            if (camera_point.at<double>(2, 0) > 0) { // Point is in front of camera
                cv::Mat image_point = camera_matrix_ * camera_point;
                double u = image_point.at<double>(0, 0) / image_point.at<double>(2, 0);
                double v = image_point.at<double>(1, 0) / image_point.at<double>(2, 0);
                
                // Check if point projects within image bounds
                if (u >= 0 && u < cv_ptr->image.cols && v >= 0 && v < cv_ptr->image.rows) {
                    cv::Vec3b color = cv_ptr->image.at<cv::Vec3b>(v, u);
                    
                    pcl::PointXYZRGB& colored_pt = colored_cloud->points[i];
                    colored_pt.x = pt.x;
                    colored_pt.y = pt.y;
                    colored_pt.z = pt.z;
                    colored_pt.r = color[2];
                    colored_pt.g = color[1];
                    colored_pt.b = color[0];
                    
                    // Draw point on debug image
                    cv::circle(debug_image, cv::Point(u, v), 2, cv::Scalar(0, 255, 0), -1);
                } else {
                    // Point outside image - use default color
                    pcl::PointXYZRGB& colored_pt = colored_cloud->points[i];
                    colored_pt.x = pt.x;
                    colored_pt.y = pt.y;
                    colored_pt.z = pt.z;
                    colored_pt.r = 128;
                    colored_pt.g = 128;
                    colored_pt.b = 128;
                }
            } else {
                // Point behind camera - use default color
                pcl::PointXYZRGB& colored_pt = colored_cloud->points[i];
                colored_pt.x = pt.x;
                colored_pt.y = pt.y;
                colored_pt.z = pt.z;
                colored_pt.r = 128;
                colored_pt.g = 128;
                colored_pt.b = 128;
            }
        }
        
        // Publish colored cloud
        sensor_msgs::PointCloud2 colored_msg;
        pcl::toROSMsg(*colored_cloud, colored_msg);
        colored_msg.header = cloud->header;
        colored_cloud_pub_.publish(colored_msg);
        
        // Publish debug image
        sensor_msgs::ImagePtr debug_msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", debug_image).toImageMsg();
        debug_msg->header = image_msg->header;
        debug_image_pub_.publish(debug_msg);
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "camera_lidar_fusion_node");
    
    CameraLidarFusion fusion_node;
    
    ros::spin();
    
    return 0;
}
